src/host/readDataCooked.cpp

```diff
@@ -436,633 +196,1043 @@ bool COOKED_READ_DATA::AtEol() const noexcept
-                                        _originalCursorPosition.x,
-                                        TRUE))
-                {
-                    if (CursorPosition.x == (sScreenBufferSizeX - 1))
-                    {
-                        CursorPosition.x++;
-                    }
-                }
-
-                // adjust cursor position for WriteChars
-                _originalCursorPosition.y += ScrollY;
-                CursorPosition.y += ScrollY;
-                AdjustCursorPosition(_screenInfo, CursorPosition, TRUE, nullptr);
+                _history->Empty();
+                _history->Flags |= CommandHistory::CLE_ALLOCATED;
+            }
+        }
+        break;
+    case VK_F8:
+        if (_history)
+        {
+            CommandHistory::Index index = 0;
+            const auto prefix = std::wstring_view{ _buffer }.substr(0, _bufferCursor);
+            if (_history->FindMatchingCommand(prefix, _history->LastDisplayed, index, CommandHistory::MatchOptions::None))
+            {
+                _buffer.assign(_history->RetrieveNth(index));
+                _bufferCursor = std::min(_bufferCursor, _buffer.size());
+                _markAsDirty();
             }
         }
+        break;
+    case VK_F9:
+        if (_history && _history->GetNumberOfCommands())
+        {
+            _popupPush(PopupKind::CommandNumber);
+        }
+        break;
+    case VK_F10:
+        // Alt+F10 clears the aliases for specifically cmd.exe.
+        if (altPressed)
+        {
+            Alias::s_ClearCmdExeAliases();
+        }
+        break;
+    default:
+        assert(false); // Unrecognized VK. Fix or don't call this function?
+        break;
     }
+}
 
-    // in cooked mode, enter (carriage return) is converted to
-    // carriage return linefeed (0xda).  carriage return is always
-    // stored at the end of the buffer.
-    if (wch == UNICODE_CARRIAGERETURN)
+// Handles any tasks that need to be completed after the read input loop finishes,
+// like handling doskey aliases and converting the input to non-UTF16.
+void COOKED_READ_DATA::_handlePostCharInputLoop(const bool isUnicode, size_t& numBytes, ULONG& controlKeyState)
+{
+    auto writer = _userBuffer;
+    std::wstring_view input{ _buffer };
+    size_t lineCount = 1;
+
+    if (WI_IsFlagSet(_pInputBuffer->InputMode, ENABLE_ECHO_INPUT))
     {
-        if (_processedInput)
+        // The last characters in line-read are a \r or \r\n unless _ctrlWakeupMask was used.
+        // Neither History nor s_MatchAndCopyAlias want to know about them.
+        const auto& suffix = _newlineSuffix();
+        if (input.ends_with(suffix))
         {
-            if (_bytesRead < _bufferSize)
+            input.remove_suffix(suffix.size());
+
+            // Reset the cursor back to its regular size if it was previously changed by our VK_INSERT handler.
+            _screenInfo.SetCursorDBMode(false);

```

Comment:
Ah whoops. So I already did the thing that Mike found, but forgot about it. After testing, I found that my new implementation (this line) is more correct than the current implementation, because pressing tab for autocompletion in cmd.exe needs to revert the cursor shape.